### PR TITLE
Reject archives containing duplicated framework names

### PR DESCRIPTION
**Pull request by blender**
See #2764 in which the archive contains "TealiumIOS.framework" multiple times.

### CODE REVIEW

File: Source/CarthageKit/Project.swift [hunk 1]
```diff
@@ -591,6 +591,16 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+					.collect()
+					// Check if the Archive contains multiple framewroks with the same `lastPathComponent`, i.e. their name.
+					// This check is needed because they will be copied over to the same destination location.
+					.flatMap(.merge) { frameworksUrls -> SignalProducer<URL, CarthageError> in
+						let setOfUniqueFrameworkNames = Set(frameworksUrls.map { $0.lastPathComponent })
+						guard setOfUniqueFrameworkNames.count == frameworksUrls.count else {
+							return SignalProducer(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same name"))
+						}
+						return SignalProducer(frameworksUrls)
+					}
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: This doesn't look like it correctly handles multiple platforms. e.g. ReactiveCocoa supports all 4 platforms, so it can contain 4 `ReactiveCocoa.framework`s.

Comment (blender) [reply Source/CarthageKit/Project.swift]: @mdiep This is why there are code reviews 😅 

It now builds the full destination path. I force pushed on amend so this comment is now misplaced.

**Comment by blender** [comment]
I'll see if I can add tests for this for the positive, false positive and negative case.

File: Source/CarthageKit/Project.swift [hunk 2]
```diff
@@ -591,6 +591,34 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+							}
+							.uniqueValues()
+							.collect()
+							.flatMap(.merge) { uniquePaths -> SignalProducer<[String], CarthageError> in
+								guard uniquePaths.count == frameworksUrls.count else {
+									return SignalProducer(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same unarchiving destination"))
+								}
+								return SignalProducer<[String], CarthageError>(value: uniquePaths)
+							}
+							.then(urlProducer)
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: This seems like it might be duplicating some of the logic below:

1. Figure out where we would copy everything
2. Make sure there's no duplicates
3. Copy everything
    1. Figure out where we would copy everything
    2. Actually copy it

Can we restructure the code to avoid that duplication?

1. Figure out what to copy where
2. Make sure there are no duplicates
3. Copy based on the list from (1)

**Comment by blender** [comment]
@mdiep I streamlined the logic so that it avoids repetition.

I don't have a good solution for testing. If I were to make it testable I would mock all the parts that do IO and collect half way results (i.e. URLs) then compare with a known lists.

Also:
- the very first `.collect` can be avoided if there were something like `uniqueValues(onceCollected:Set<U> -> Bool)`
- flatMapping to do a side effect and forward the same producer is pretty annoying. Is there a better way?

File: Source/CarthageKit/Project.swift [hunk 3]
```diff
@@ -587,21 +587,65 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+										return .success(self
+											.directoryURL
+											.appendingPathComponent(destinationURLInWorkingDir.path, isDirectory: true))
+								}
+							}
+							.uniqueValues()
+							.collect()
+							.flatMap(.merge) { uniqueDestinationUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+								guard uniqueDestinationUrls.count == frameworksUrls.count else {
+									return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same unarchiving destination"))
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: Ideally this error would say _which_ frameworks have the same destination

File: Source/CarthageKit/Project.swift [hunk 4]
```diff
@@ -587,21 +587,65 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+								}
+
+								let frameworkUrlAndDestinationUrlPairs = Array(zip(frameworksUrls, uniqueDestinationUrls)
+									.map { FrameworkUrlAndDestinationUrl(frameworkUrlInUnarchivedFolder:$0, frameworkUrlInBuildForlder: $1) })
+								return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(frameworkUrlAndDestinationUrlPairs)
+							}
+					}
+					// Check if the framework are compatible with the current Swift version
+					.flatMap(.merge) { pair -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+						return checkFrameworkCompatibility(pair.frameworkUrlInUnarchivedFolder, usingToolchain: toolchain)
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: Should we do this filter before the above error? Then you could bundle multiple Swift versions? Or maybe this is unimportant with ABI stability.

**Comment by mdiep** [comment]
Realistically, I think there are 2 options for making this testable:

- Mocking with dependency injection. I've often used default arguments to make this easy. (So the really caller doesn't need to pass in any dependencies.)

- Separate the pure parts from the impure parts and test the pure parts.

Neither seems like a great option here, so I'll leave it to your discretion. :)

File: Source/CarthageKit/Project.swift (hunk 4 above)
Comment (blender) [reply Source/CarthageKit/Project.swift]: If we picked only the swift version of the framework base on a smarter compatibility check then yes, the compatibility check should happen before. I can move it now, however the compatibility takes longer than checking for duplicate destination paths

File: Source/CarthageKit/Project.swift (hunk 3 above)
Comment (blender) [reply Source/CarthageKit/Project.swift]: Yeah... I was lazy 😅  I'll try to make it smarter

**Comment by blender** [comment]
I have prettified the error message to:

```
Invalid archive - Found multiple frameworks with the same unarchiving destination:
* 	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/TealiumIOSLifecycle.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/DevicesOnly/TealiumIOSLifecycle.framework/ 
		to:
	file:///Users/Tommaso/Code/lab/Carthage-2764/Carthage/Build/iOS/TealiumIOSLifecycle.framework/
* 	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/Samples/ConsentManagerDemo_ObjC/ConsentManagerDemo_ObjC/Base.lproj/TealiumIOS.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/Samples/ConsentManagerDemo_Swift/ConsentManagerDemo_Swift/Base.lproj/TealiumIOS.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/DevicesOnly/TealiumIOS.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/TealiumIOS.framework/ 
		to:
	file:///Users/Tommaso/Code/lab/Carthage-2764/Carthage/Build/iOS/TealiumIOS.framework/
* 	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/TealiumIOSOptimizely.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/DevicesOnly/TealiumIOSOptimizely.framework/ 
		to:
	file:///Users/Tommaso/Code/lab/Carthage-2764/Carthage/Build/iOS/TealiumIOSOptimizely.framework/
* 	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/TealiumWATCHOSExtension.framework/
	file:///var/folders/b4/qz7m1r9x3mj1bm3nfsjjf64m0000gp/T/carthage-archive.Fyy5fV/tealium-ios-5.4.11/DevicesOnly/TealiumWATCHOSExtension.framework/ 
		to:
	file:///Users/Tommaso/Code/lab/Carthage-2764/Carthage/Build/watchOS/TealiumWATCHOSExtension.framework/
```

File: Source/CarthageKit/Project.swift [hunk 5]
```diff
@@ -587,21 +587,81 @@ public final class Project { // swiftlint:disable:this type_body_length
 		pinnedVersion: PinnedVersion,
 		toolchain: String?
 	) -> SignalProducer<URL, CarthageError> {
+		
+		typealias FrameworkUrlAndDestinationUrl = (frameworkUrlInUnarchivedFolder: URL, frameworkUrlInBuildForlder: URL)
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: _URL_ should be all caps.

_Folder_ is misspelled in the second label.

Comment (mdiep) [reply Source/CarthageKit/Project.swift]: Would it be better to call these _sourceURL_ and _destinationURL_?

File: Source/CarthageKit/Project.swift [hunk 6]
```diff
@@ -587,21 +587,81 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+										guard let destinationURLInWorkingDir = platform
+											.relativeURL?
+											.appendingPathComponent(url.lastPathComponent, isDirectory: true) else {
+												return .failure(.internalError(description: "failed to construct framework destination url from \(platform) and \(url.lastPathComponent)"))
+										}
+
+										return .success(self
+											.directoryURL
+											.appendingPathComponent(destinationURLInWorkingDir.path, isDirectory: true))
+								}
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: Would it make sense to extract the contents of this `flatMap` to a helper function? `destinationURLForFramework(at: url)`?

File: Source/CarthageKit/Project.swift [hunk 7]
```diff
@@ -587,21 +587,81 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+
+									return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(error: CarthageError.duplicatesInArchive(duplicates: dupes))
+								}
+
+								let uniquePairs = destinationMap
+									.filter { $0.value.count == 1}
+									.map { FrameworkUrlAndDestinationUrl(frameworkUrlInUnarchivedFolder:$0.key,
+																		 frameworkUrlInBuildForlder: $0.value.first!)}
+
+								return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(uniquePairs)
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: Same here. Could this be extracted to a helper function?

I find the code much easier to follow when the `flatMap`s are very high level.

**Comment by mdiep** [comment]
This looks good! I mostly want the typo to be fixed. ☺️

**Comment by blender** [comment]
I have extracted some functions out of the flatMaps but to extract the last one, I would have to expose 2 types that are really only used in one place. I don't know if this makes for more readable code or not.

Both functions I made are pure so those could be tested but I would rather keep them private and just mock stuff in the exposed function

File: Source/CarthageKit/Project.swift (hunk 6 above)
Comment (blender) [reply Source/CarthageKit/Project.swift]: done, slightly different name

File: Source/CarthageKit/Project.swift (hunk 7 above)
Comment (blender) [reply Source/CarthageKit/Project.swift]: I only moved it out of the flatMap but it's still inside the function. The purpose of the function is really to `uniquePairs` keyed by the first element of the pair.  I didn't make it super generic because I still kept the context, but I could. However I need a `GenericError` since result is not Either and only takes `: Error`.

Something like

```
		func uniquePairs<A, B>(
			_ pairs: [(A, B)]
			) -> Result<[(A, B>], GenericError<[A: [B]]>> {
			let map = pairs
				.reduce([A: [B]]()) { dict, pair in
					var mutableDict = dict
					mutableDict[pair.0] =
						(mutableDict[pair.0] ?? []) + [pair.1]

					return mutableDict
			}

			let dupes = map.filter { $0.value.count > 1 }
			guard dupes.count == 0 else {
				return .failure(GenericError<[A: [B]]>(value: dupes))
			}

			let uniquePairs = map
				.filter { $0.value.count == 1}
				.map { ($0, $0.value.first!)}
			return .success(uniquePairs)
		}
```

File: Source/CarthageKit/Errors.swift [hunk 8]
```diff
@@ -185,6 +189,8 @@ extension CarthageError: Equatable {
 		case let (.internalError(left), .internalError(right)):
 			return left == right
 
+		case let (.duplicatesInArchive(left), .duplicatesInArchive(right)):
+			return left == right
 		default:
```
Comment (mdiep) [inline Source/CarthageKit/Errors.swift]: There should be an empty line between these.

File: Source/CarthageKit/Project.swift [hunk 9]
```diff
@@ -587,21 +610,89 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+			let value: A
+		}
+
+		// Returns the unique pairs in the input array
+		// or the duplicates keys by .frameworkDestinationURL
+		func uniqueSourceDestinationPairs(
+			_ sourceURLAndDestinationURLpairs: [SourceURLAndDestinationURL]
+			) -> Result<[SourceURLAndDestinationURL], GenericError<[URL: [URL]]>> {
+			let destinationMap = sourceURLAndDestinationURLpairs
+				.reduce([URL: [URL]]()) { dict, pair in
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: This would be simpler with `reduce(into:)`

File: Source/CarthageKit/Project.swift [hunk 10]
```diff
@@ -587,21 +610,89 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+
 		return SignalProducer<URL, CarthageError>(value: zipFile)
 			.flatMap(.concat, unarchive(archive:))
 			.flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
+				// For all frameworks in the directory where the archive has been expanded
 				return frameworksInDirectory(directoryURL)
-					.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
-						return checkFrameworkCompatibility(url, usingToolchain: toolchain)
+					.collect()
+					// Check if multiple framewroks resolve to the same unique destination URL in the Carhage/Build/ folder.
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: Typo of `frameworks` and `Carthage` here 🙈

File: Source/CarthageKit/Project.swift [hunk 11]
```diff
@@ -587,21 +610,89 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+		// Helper type
+		struct GenericError<A>: Error {
+			let value: A
+		}
+
+		// Returns the unique pairs in the input array
+		// or the duplicates keys by .frameworkDestinationURL
+		func uniqueSourceDestinationPairs(
+			_ sourceURLAndDestinationURLpairs: [SourceURLAndDestinationURL]
+			) -> Result<[SourceURLAndDestinationURL], GenericError<[URL: [URL]]>> {
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: I would either:

1. Return a `CarthageError`

2. Add a `DuplicatesInArchiveError` type and make `CarthageError`'s case `case duplicatesInArchive(DuplicatesInArchiveError)`

(I'd prefer to move towards the second option, personally)

File: Source/CarthageKit/Project.swift [hunk 12]
```diff
@@ -587,21 +610,89 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+									.flatMap(.merge) { SignalProducer($0) }
+							}
+					}
+					// Check if the framework are compatible with the current Swift version
+					.flatMap(.merge) { pair -> SignalProducer<SourceURLAndDestinationURL, CarthageError> in
+						return checkFrameworkCompatibility(pair.frameworkSourceURL, usingToolchain: toolchain)
 							.mapError { error in CarthageError.internalError(description: error.description) }
+							.then(SignalProducer<SourceURLAndDestinationURL, CarthageError>(value: pair))
+					}
+					// If the framework is compatible copy it over to the destination folder in Carhage/Build
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: typo: `Carthage`

**Comment by mdiep** [comment]
Again, I mostly want to fix the typos. 😅

File: Source/CarthageKit/Errors.swift [hunk 13]
```diff
@@ -8,6 +8,11 @@ import XCDBLD
 public enum CarthageError: Error {
 	public typealias VersionRequirement = (specifier: VersionSpecifier, fromDependency: Dependency?)
 
+	public struct DuplicatesInArchive: Equatable {
```
Comment (blender) [inline Source/CarthageKit/Errors.swift]: @mdiep done

**Comment by blender** [comment]
@mdiep comments and typos (🙈 ) addressed. Squashed everything.

**Comment by ffittschen** [comment]
Hi @blender and @mdiep, we just tried out a locally built version of carthage (built from the `master` branch that contains this PR) in order to see if it fixed #2777.

However, it seems like this PR introduced some other issue where the error message reported in #2777 is not appearing anymore, but the whole `iOS` folder in `Carthage/Build` is missing! It only creates the `.version` files with an empty `iOS` array.

<details>

<summary>Cartfile</summary>
```
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseRemoteConfigBinary.json"
```
</details>

<details>

<summary>Log output</summary>
```
$ carthage bootstrap --platform ios

*** Downloading binary-only framework FirebaseAnalyticsBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
*** xcodebuild output can be found in /var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-xcodebuild.1QLaZ9.log
*** Downloading binary-only framework FirebaseMessagingBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json"
*** Downloading binary-only framework FirebaseRemoteConfigBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseRemoteConfigBinary.json"
*** Downloading binary-only framework GoogleSymbolUtilities at "file:///Users/ffittschen/Code/demo-project/Carthage/BinaryFrameworkSchemes/GoogleSymbolUtilities.json"
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.fvIclQ/Frameworks/frameworks/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.INU35V/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.INU35V/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.INU35V/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.oXnQFW/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.oXnQFW/
*** Downloading binary-only framework GoogleTagManager at "file:///Users/ffittschen/Code/demo-project/Carthage/BinaryFrameworkSchemes/GoogleTagManager.json"
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.DrleZQ/Frameworks/
*** Downloading binary-only framework GoogleUtilitiesLegacy at "file:///Users/ffittschen/Code/demo-project/Carthage/BinaryFrameworkSchemes/GoogleUtilitiesLegacy.json"
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.RJwYOS/Frameworks/frameworks/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
file:///var/folders/m2/8kggf73s7tdc3c14z5btfx5w0000gn/T/carthage-archive.eA8C6K/
```
</details>

**Comment by blender** [comment]
Checking

**Comment by blender** [comment]
Confirmed

**Comment by blender** [comment]
Fix in #2791
